pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/operands/FilterOperand.java

```diff
@@ -230,14 +230,41 @@ public Predicate(List<RexExpression> functionOperands, DataSchema dataSchema) {
           "Expected 2 function ops for Predicate but got:" + functionOperands.size());
       _lhs = TransformOperand.toTransformOperand(functionOperands.get(0), dataSchema);
       _rhs = TransformOperand.toTransformOperand(functionOperands.get(1), dataSchema);
-      if (_lhs._resultType != null && _lhs._resultType != DataSchema.ColumnDataType.OBJECT) {
-        _resultType = _lhs._resultType;
-      } else if (_rhs._resultType != null && _rhs._resultType != DataSchema.ColumnDataType.OBJECT) {
-        _resultType = _rhs._resultType;
+      _resultType = resolveResultType(_lhs._resultType, _rhs._resultType);
+    }
+
+    private static DataSchema.ColumnDataType resolveResultType(DataSchema.ColumnDataType lhsType,
+        DataSchema.ColumnDataType rhsType) {
+      if (lhsType == null) {
+        return rhsType;
+      } else if (rhsType == null) {
+        return lhsType;
       } else {
-        // TODO: we should correctly throw exception here. Currently exception thrown during constructor is not
-        // piped back to query dispatcher, thus we set it to null and deliberately make the processing throw exception.
-        _resultType = null;
+        if (isSuperType(lhsType, rhsType)) {
+          return lhsType;
+        } else if (isSuperType(rhsType, lhsType)) {
+          return rhsType;
+        } else {
+          return null;
+        }
+      }
+    }
+
+    private static boolean isSuperType(DataSchema.ColumnDataType superType, DataSchema.ColumnDataType subType) {
```

Comment:
hmm, this seems pretty hacky, I think this solution needs a bit of tweaking (I think it handles Objects in the opposite way you intended) but what do you think of this:
```java
  public static void main(String[] args) {
    foo(ColumnDataType.LONG, ColumnDataType.INT);
    foo(ColumnDataType.STRING, ColumnDataType.INT);
    foo(ColumnDataType.OBJECT, ColumnDataType.STRING);
    foo(ColumnDataType.LONG, ColumnDataType.DOUBLE);
  }

  static void foo(ColumnDataType lType, ColumnDataType rType) {
    Class<?> lClazz = Optional.ofNullable(lType.getNullPlaceholder()).orElse(new Object()).getClass();
    Class<?> rClazz = Optional.ofNullable(rType.getNullPlaceholder()).orElse(new Object()).getClass();

    Ordering<Class<?>> numerics = Ordering.explicit(
        Integer.class,
        Long.class,
        Float.class,
        Double.class,
        BigDecimal.class
    );

    if (Number.class.isAssignableFrom(lClazz) && Number.class.isAssignableFrom(rClazz)) {
      System.out.println(numerics.max(lClazz, rClazz));
    } else if (lClazz.isAssignableFrom(rClazz)) {
      System.out.println(lClazz);
    } else if (rClazz.isAssignableFrom(lClazz)) {
      System.out.println(rClazz);
    } else {
      System.out.println("null");
    }
  }
```

Reply:
- actually good idea using `getNullPlaceholder()` since it already has the java class.
- Object handling should be correct but I might be wrong as I haven't add a unit test, yet

Reply:
incorporated this thought, plz kindly take another look